test/UpgradeableProxyTest.js

```diff
@@ -0,0 +1,65 @@
+// JS Libraries
+const withData = require("leche").withData;
+const { t } = require("./utils/consts");
+
+// Smart contracts
+const AdminUpgradeabilityProxy = artifacts.require("./base/UpgradeableProxy.sol");
+const UpgradableV1 = artifacts.require("./mock/upgradable/UpgradableV1.sol");
+const UpgradableV2 = artifacts.require("./mock/upgradable/UpgradableV2.sol");
+
+const Mock = artifacts.require("./mock/util/Mock.sol");
+
+contract("UpgradeableProxyTest", function(accounts) {
+    let v1LibraryInstance;
+    let v2LibraryInstance;
+    let v1InitData;
+    let initValue = 43;
+
+    beforeEach("Setup for each test", async () => {
+        v1LibraryInstance = await UpgradableV1.new();
+        v2LibraryInstance = await UpgradableV2.new();
+
+        v1InitData = v1LibraryInstance.contract.methods.initialize(initValue).encodeABI();
+    });
+
+    withData({
+        _1_only_admin_upgrade: [ accounts[1], accounts[0], false, true, "UPGRADABLE_CALLER_MUST_BE_ADMIN" ],
+        _2_successful: [ accounts[0], accounts[0], false, false, null ],
+        _3_initialize_after_constructor: [ accounts[0], accounts[0], true, false, null ]
+    }, function(caller, admin, initAfter, mustFail, expectedErrorMessage) {
+        it(t("user", "transfer", "Should be able transfer tokens.", mustFail), async function() {
+            try {
+                // Setup
+                const initData = initAfter ? "0x" : v1InitData;
+                const proxy = await AdminUpgradeabilityProxy.new(v1LibraryInstance.address, admin, initData);
+                const v1 = await UpgradableV1.at(proxy.address);
+                const v2 = await UpgradableV2.at(proxy.address);
+
+                // Invocation
+                if (initAfter) {
+                    await v1.initialize(initValue);
+                }
+
+                const implementationV1 = await proxy.implementation.call();
+                assert.equal(implementationV1.toLowerCase(), v1LibraryInstance.address.toLowerCase(), "V1 implementation addresses do not match");
+
+                const v1Value = await v1.value.call();
+                assert.equal(initValue.toString(), v1Value.toString(), "Initial values do not match");
+
+                await proxy.upgradeTo(v2LibraryInstance.address, { from: caller });
+                const implementationV2 = await proxy.implementation.call();
+                assert.equal(implementationV2.toLowerCase(), v2LibraryInstance.address.toLowerCase(), "V2 implementation addresses do not match");
+
+                const newValue = 1234;
+                await v2.setValue(newValue);
```

Comment:
Regarding the ether/erc20 balances. How this proxy deal with it? I think we should have unit tests to verify it.

Reply:
I'll create a unit test for that. But they won't change. because the balances will be on the `proxy` address which is the only address that we need to know in order to interact with the contract. It just forwards any call to the current `implementation` contract to run the current version of functionality.

Reply:
It would be great if we could have a test to validate we can freely access/transfer eth from both versions V1 and V2. Let's say accumulate eth using V1 (should remain in proxy), then upgrade to V2 and transfer some of those funds.

Reply:
I can definitely create a test to demonstrate this but the funds will only be interacting with the `proxy.address`. Notice these lines: https://github.com/teller-protocol/teller-protocol-v1/pull/118/files/6ce23521c0e7a4ab608687abc9d5fd504ff0ca74#diff-b66228bea33641cbcf8a0c2f1387bcb7R35-R36